python/ml4ir/config/features.py

```diff
@@ -1,120 +1,237 @@
+                                     # Defaults to just padding filling null values
+        required: <bool> # Whether the feature is mandatory at serving time
+    tfrecord_type: <context or sequence | str>
+rank:  # Field representing the initial rank of the record in a query
+    ...
+label:  # Binary click label
     ...
-}
+features:
+    - name: feature_0
+        ...
+    - name: feature_1
+        ...
+    ...
+
 """
 
 
-class Features:
+class FeatureConfigKey:
+    QUERY_KEY = "query_key"
+    RANK = "rank"
+    LABEL = "label"
+    FEATURES = "features"
+
+
+class FeatureConfig:
     """
     Class to store features to be used for the ranking model
     """
 
-    def __init__(self, feature_config: dict):
-        self.feature_config = feature_config
-
-        self.record_features: List[str] = self.get_feature_names(trainable=True)
-        self.metadata_features: List[str] = self.get_feature_names(trainable=False)
-        # TODO: Define query level features and clean up tfrecord data format
-        self.query_features = None
-        self.query_key = self.get_query_key()
-        # TODO: this loop is silly - the label should be trackable without looping over the whole dict.
-        for feature, feature_info in self.feature_config.items():
-            if feature_info["type"] == "label":
-                self.label = feature
-        self.features_to_log = {"new_score", "new_pos"}
-        for feature, feature_info in self.feature_config.items():
-            if feature_info.get("log", False):
-                self.features_to_log.add(feature)
-        if len(self.record_features) == 0:
+    def __init__(self, features_dict, logger: Optional[Logger] = None):
+        self.all_features = list()
+
+        try:
+            self.query_key = features_dict.get(FeatureConfigKey.QUERY_KEY)
+            self.all_features.append(self.query_key)
+        except KeyError:
+            self.query_key = None
+            if logger:
+                logger.warning("'query_key' key not found in the feature_config specified")
+
+        try:
+            self.rank = features_dict.get(FeatureConfigKey.RANK)
+            self.all_features.append(self.rank)
+        except KeyError:
+            self.rank = None
+            if logger:
+                logger.warning("'rank' key not found in the feature_config specified")
+
+        try:
+            self.label = features_dict.get(FeatureConfigKey.LABEL)
+            self.all_features.append(self.label)
+        except KeyError:
+            raise KeyError("'label' key not found in the feature_config specified")
+
+        try:
+            self.features: List = features_dict.get(FeatureConfigKey.FEATURES)
+            self.all_features.extend(self.features)
+        except KeyError:
+            raise KeyError("'features' key not found in the feature_config specified")
+
+        # Features that can be used for training the model
+        self.ranking_features = list()
```

Comment:
what type of list is this?  can we have types for these?  `List[str]` or whatnot?

Reply:
I initially had it as `List[Dict[str, dict]]`, but then everything blew up and led to the `35` errors I posted above. I'm not sure how I would define the type for this effectively without breaking anything downstream.

The list essentially contains the feature info dictionaries for each feature.

Look at `get_ranking_features()`. It makes it even worse to define a dictionary because that method could essentially return anything(value in the dictionary).

If you want me to add the type for **only** this list, I could may be do that though.

Reply:
oh I see recursive type is pretty nested.  Maybe a struct class to handle this config, instead of basically just JSON in memory?  But not for now I guess.   But something to consider...